### PR TITLE
Change how chain links are stored

### LINKED ISSUE #480: Change how chain links are stored
## Overview
When working towards #472, I realized that currently chain links are stored associated with a profile inside the `Profile` object itself. While this is very handy when querying them, the major downside is that when storing a chain link we need to update the **whole** profile. Since the gas usage is based on how much data is written during a transaction, this will cause later changes to the chain links slice to be extremely costly. Suppose we have a profile with already 10 links, if we add the 11th we need to write the whole slice again so we are writing 10 objects that we should not. 

## Changes proposal
For the above reason, I think what we can do is change how chain links are stored by using the same method we use with other repeated data (eg: DTag requests, blocks, etc): 

```
ChainLinkPrefix + username + chain_name + chain_address -> ChainLink
```

This would allow us to check the existence of a link in O(1) time as well as to query for such link in the same time. 

If this is implemented, the `ProfileByChainLinkQuery` needs to be replaced with a `UserChainLinks` **paginated** query.

Please @dadamu @bragaz let me know what you think about this change.

### PR DESCRIPTION
**Pull request by dadamu**
## Description

This PR changes how chain links are stored and allows to query paginated chain links.
Close #480

## Checklist
- [x] Targeted PR against correct branch.
- [x] Linked to Github issue with discussion and accepted design OR link to spec that describes this work.
- [x] Wrote unit tests.
- [ ] Wrote integration tests (simulation & CLI). 
- [ ] Updated the documentation. 
- [x] Added an entry to the `CHANGELOG.md` file.
- [ ] Re-reviewed `Files changed` in the Github PR explorer.

### CODE REVIEW

**Comment by RiccardoM** [comment]
Instead of defining a `ChainLinkEntry`, what I think we can do is simply add the `User` field to the `ChainLink` message itself. This should make it more easy to manage, export and import.

File: x/profiles/keeper/keeper_chain_links.go [hunk 1]
```diff
@@ -47,60 +48,44 @@ func (k Keeper) StoreChainLink(ctx sdk.Context, user string, link types.ChainLin
...
+		)
+	}
+	store.Delete(types.ChainLinksStoreKey(owner, chainName, target))
 	return nil
 }
 
-// GetAccountByChainLink returns the account corresponding to the given address and the given chain name
-func (k Keeper) GetAccountByChainLink(ctx sdk.Context, chainName string, address string) (sdk.AccAddress, bool) {
-	key := types.ChainsLinksStoreKey(chainName, address)
+func (k Keeper) GetChainLink(ctx sdk.Context, owner, chainName, target string) (types.ChainLink, bool) {
```
[the remark this answers is not in the document]
Comment (dadamu) [reply x/profiles/keeper/keeper_chain_links.go:72]: @bragaz There is a reason that I would avoid returning small struct pointer. As far as I know, Returning pointer would increase pressure on garbage collection. Returning value would be more faster than pointer.

Reference:
https://medium.com/a-journey-with-go/go-should-i-use-a-pointer-instead-of-a-copy-of-my-struct-44b43b104963

[the remark this answers is not in the document]
Comment (bragaz) [reply x/profiles/keeper/keeper_chain_links.go:72]: @dadamu Didn't know this! All right then! Thanks for the ref, interesting channel to follow

File: x/profiles/keeper/alias_functions.go [hunk 2]
```diff
@@ -75,3 +75,49 @@ func (k Keeper) IterateUserRelationships(ctx sdk.Context, user string, fn func(i
...
+		link := types.MustUnmarshalChainLink(k.cdc, iterator.Value())
+
+		stop := fn(i, link)
+
+		if stop {
+			break
+		}
+
+		i++
+	}
```
Comment (RiccardoM) [inline x/profiles/keeper/alias_functions.go:100]: Just some small changes to remove empty lines
```suggestion
	i := int64(0)
	for ; iterator.Valid(); iterator.Next() {
		link := types.MustUnmarshalChainLink(k.cdc, iterator.Value())

		stop := fn(i, link)
		if stop {
			break
		}
		i++
	}
```

File: x/profiles/keeper/alias_functions.go [hunk 3]
```diff
@@ -75,3 +75,49 @@ func (k Keeper) IterateUserRelationships(ctx sdk.Context, user string, fn func(i
...
```
Comment (RiccardoM) [inline x/profiles/keeper/alias_functions.go:122]:
```suggestion
	i := int64(0)
	for ; iterator.Valid(); iterator.Next() {
		link := types.MustUnmarshalChainLink(k.cdc, iterator.Value())

		stop := fn(i, link)
		if stop {
			break
		}
		i++
	}
```

**Review by RiccardoM — COMMENTED** [review]
Just a couple of minor improvements

**Review by RiccardoM — APPROVED** [review]

**Pull request merged**